### PR TITLE
Add Device table

### PR DESCRIPTION
**Pull request by gdong1**
Update core code to support device table.
Updated boot option to consume device table.
Update firmware update to consume device table.
Update shell command on boot option changes.
Add device table for APL and Qemu.
Remove unnecessary code.

Signed-off-by: Guo Dong <guo.dong@intel.com>

### CODE REVIEW

File: BootloaderCommonPkg/Include/Guid/DeviceTableHobGuid.h [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+
+#ifndef __DEVICE_TABLE_HOB_GUID_H__
+#define __DEVICE_TABLE_HOB_GUID_H__
+
+///
+/// Device table Hob GUID
+///
+extern EFI_GUID gDeviceTableHobGuid;
+
+typedef struct {
```
Comment (myguitar) [inline BootloaderCommonPkg/Include/Guid/DeviceTableHobGuid.h]: typedef struct {
  UINT32       PciFunctionNumber: 8;
  UINT32       PciDeviceNumber: 8;
  UINT32       PciBusNumber: 8;
  UINT32       IsMmioDevice: 8;
}
or
typedef struct {
  UINT32       Rsvd: 12;
  UINT32       PciFunctionNumber: 3;
  UINT32       PciDeviceNumber: 5;
  UINT32       PciBusNumber: 4;
  UINT32       IsMmioDevice: 8;
}

File: BootloaderCommonPkg/Include/Guid/DeviceTableHobGuid.h [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+**/
+
+#ifndef __DEVICE_TABLE_HOB_GUID_H__
+#define __DEVICE_TABLE_HOB_GUID_H__
+
+///
+/// Device table Hob GUID
+///
+extern EFI_GUID gDeviceTableHobGuid;
+
```
Comment (myguitar) [inline BootloaderCommonPkg/Include/Guid/DeviceTableHobGuid.h]: #pragma pack(1)

File: BootloaderCommonPkg/Include/Guid/DeviceTableHobGuid.h [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+  } Dev;
+} SBL_DEVICE;
+
+#define    SBL_DEVICE_TABLE_SIGNATURE  0xD11D
+typedef struct {
+  UINT16          Signature;
+  UINT16          DeviceNumber;
+  SBL_DEVICE      Device[0];
+} SBL_DEVICE_TABLE;
+
```
Comment (myguitar) [inline BootloaderCommonPkg/Include/Guid/DeviceTableHobGuid.h]: #pragma pack()

File: Platform/ApollolakeBoardPkg/Library/Stage1BBoardInitLib/Stage1BBoardInitLib.c [hunk 4]
```diff
@@ -56,6 +57,17 @@
 
 #define MRC_PARAMS_BYTE_OFFSET_MRC_VERSION 14
```
Comment (myguitar) [inline Platform/ApollolakeBoardPkg/Library/Stage1BBoardInitLib/Stage1BBoardInitLib.c]: CONST

File: Platform/ApollolakeBoardPkg/Library/Stage1BBoardInitLib/Stage1BBoardInitLib.c [hunk 5]
```diff
@@ -56,6 +57,17 @@
...
 
+SBL_DEVICE  mSblDevices[]= {
+  // Device type     instance  rsvd   fun  dev  bus  ismmio
+  {OsBootDeviceSata  , 0,       0,   {{ 0,  18,   0,    0 }} },
```
Comment (myguitar) [inline Platform/ApollolakeBoardPkg/Library/Stage1BBoardInitLib/Stage1BBoardInitLib.c]: why don't you use a macro for each device info?
ex) PCI_DEVICE(b, d, f), ADDR_DEVICE(addr)

File: BootloaderCorePkg/Stage1B/Stage1B.c [hunk 6]
```diff
@@ -463,6 +464,14 @@ SecStartup2 (
     }
   }
 
+  // Copy device table to memory
+  if (LdrGlobal->DeviceTable != NULL) {
+    DeviceTable = (SBL_DEVICE_TABLE *) LdrGlobal->DeviceTable;
+    AllocateLen = DeviceTable->DeviceNumber * sizeof (SBL_DEVICE) + sizeof (SBL_DEVICE_TABLE);
+    LdrGlobal->DeviceTable = AllocatePool (AllocateLen);
```
Comment (myguitar) [inline BootloaderCorePkg/Stage1B/Stage1B.c]: Do we need to allocate memory if DeviceNumber is 0?

File: BootloaderCorePkg/Stage2/Stage2Support.c [hunk 7]
```diff
@@ -607,16 +609,12 @@ BuildExtraInfoHob (
...
-  PerformanceInfo = BuildGuidHob (&gLoaderPerformanceInfoGuid, Length);
-  if (PerformanceInfo != NULL) {
-    PerformanceInfo->Revision = 1;
-    PerformanceInfo->Count = (UINT16)LdrGlobal->PerfData.PerfIndex;
-    PerformanceInfo->Frequency = LdrGlobal->PerfData.FreqKhz;
-    PerformanceInfo->Flags = 0;
-    CopyMem (PerformanceInfo->TimeStamp, LdrGlobal->PerfData.TimeStamp, sizeof (UINT64) * Count);
+  // Build device table Hob
+  DeviceTable  = (SBL_DEVICE_TABLE *)LdrGlobal->DeviceTable;
+  Length = sizeof (SBL_DEVICE_TABLE) + sizeof (SBL_DEVICE) * DeviceTable->DeviceNumber;
```
Comment (myguitar) [inline BootloaderCorePkg/Stage2/Stage2Support.c]: Do we need to create a hob if DeviceNumber is 0?

File: BootloaderCorePkg/Stage2/Stage2Support.c [hunk 8]
```diff
@@ -637,6 +635,18 @@ BuildExtraInfoHob (
     PlatformUpdateHobInfo (&gLoaderPlatformInfoGuid, LoaderPlatformInfo);
   }
 
+  // Build Performance Hob
```
Comment (myguitar) [inline BootloaderCorePkg/Stage2/Stage2Support.c]: Different topic - this can be in a separate PR.

File: PayloadPkg/Library/PayloadLib/PayloadLib.c [hunk 9]
```diff
@@ -360,3 +360,62 @@ GetLoaderStage (
 {
   return LOADER_STAGE_PAYLOAD;
 }
+
+/**
+  Get device address
+
+  @param[in]  DeviceType         The device type, refer OS_BOOT_MEDIUM_TYPE.
+  @param[in]  DeviceInstance     The device instance number starting from 0.
+  @param[in]  PciMmioAddr        Only applicable to PCI device. TRUE will return
```
Comment (mauricema) [inline PayloadPkg/Library/PayloadLib/PayloadLib.c]: Is PciMmioAddr useful ?   Why not always return the direct device address in the table.   If they need the PCIE MMIO address, they can always use macro to convert it.  I would recommend to make interface simpler.

File: BootloaderCommonPkg/Include/Guid/DeviceTableHobGuid.h [hunk 10]
```diff
@@ -0,0 +1,47 @@
...
+  UINT8         Type;
+  UINT8         Instance;
+  UINT16        Reserved;
+  union {
+    SBL_PCI_DEVICE  PciDev;
+    UINT32          DevAddr;
+  } Dev;
+} SBL_DEVICE;
+
+#define    SBL_DEVICE_TABLE_SIGNATURE  0xD11D
```
Comment (mauricema) [inline BootloaderCommonPkg/Include/Guid/DeviceTableHobGuid.h]: For signature,  better to use UINT32.  And better to make it readable so that it can easily identified.

File: BootloaderCorePkg/Library/BootloaderLib/BootloaderLib.c [hunk 11]
```diff
@@ -352,3 +352,55 @@ GetFeatureCfg (
...
+    DeviceBase = (UINTN)MM_PCI_ADDRESS (
+                          Device->Dev.PciDev.PciBusNumber,
+                          Device->Dev.PciDev.PciDeviceNumber,
+                          Device->Dev.PciDev.PciFunctionNumber,
+                          0);
+  } else {
+    DeviceBase = Device->Dev.DevAddr;
+  }
+
+  return DeviceBase;
```
Comment (mauricema) [inline BootloaderCorePkg/Library/BootloaderLib/BootloaderLib.c]: can we always return Device->Dev.DevAddr ?

File: PayloadPkg/Library/PayloadLib/PayloadLib.c [hunk 12]
```diff
@@ -360,3 +360,62 @@ GetLoaderStage (
...
+                                 PCI MMIO address, else return PCI BDF.
+
+  @retval     Device address for a given device instance, return 0 if the device
+              could not be found from device table.
+**/
+UINT32
+EFIAPI
+GetDeviceAddr (
```
Comment (mauricema) [inline PayloadPkg/Library/PayloadLib/PayloadLib.c]: Can we share the code with Stage ?   Maybe move it to common.   Only need to implement GetDeviceTable() API differently ?

File: PayloadPkg/Library/SblParameterLib/SblParameterLib.c [hunk 13]
```diff
@@ -327,7 +329,8 @@ AddSblCommandLine (
   AsciiSPrint (ParamValue, sizeof (ParamValue), " bdev=%a", GetBootDeviceNameString (BootOption->DevType));
   AsciiStrCatS (CommandLine, MaxCmdSize, ParamValue);
 
-  AsciiSPrint (ParamValue, sizeof (ParamValue), " diskbus=0x%x", BootOption->DevAddr);
+  DiskBus = GetDeviceAddr (BootOption->DevType, BootOption->DevInstance, FALSE);
```
Comment (mauricema) [inline PayloadPkg/Library/SblParameterLib/SblParameterLib.c]: What is disk bus ?  I  thought the value returned is device MMIO address.

File: Platform/QemuBoardPkg/Library/Stage1BBoardInitLib/Stage1BBoardInitLib.c [hunk 14]
```diff
@@ -196,9 +206,15 @@ BoardInit (
 )
 {
   EFI_STATUS            Status;
+  SBL_DEVICE_TABLE      *SblDeviceTable;
 
   switch (InitPhase) {
   case PreConfigInit:
+    SblDeviceTable = (SBL_DEVICE_TABLE *)AllocatePool (sizeof (SBL_DEVICE_TABLE) + sizeof (mSblDevices));
```
Comment (mauricema) [inline Platform/QemuBoardPkg/Library/Stage1BBoardInitLib/Stage1BBoardInitLib.c]: can we use the same mechnism as other buffer, such as CfgData,  FlashMap, etc ?  In this way, it is consistent and easier to optimize later on.

Comment (gdong1) [reply Platform/QemuBoardPkg/Library/Stage1BBoardInitLib/Stage1BBoardInitLib.c]: CfgData and FlashMap space will be available in Stage1A phase, it means it will use CAR memory if we uses same way. Just want to avoid to use CAR memory unless it is necessary since CAR memory is rare in size.

File: BootloaderCommonPkg/Include/Guid/DeviceTableHobGuid.h (hunk 2 above)
Comment (gdong1) [reply BootloaderCommonPkg/Include/Guid/DeviceTableHobGuid.h]: It is not required since the structure is 4 bytes aligned in purpose. and it is the preferred to avoid to use pack()

File: BootloaderCommonPkg/Include/Guid/DeviceTableHobGuid.h (hunk 1 above)
Comment (gdong1) [reply BootloaderCommonPkg/Include/Guid/DeviceTableHobGuid.h]: basically the first one is same with my definition. and if you want to access whole struct, you could use DevAddr since it is a union.

File: BootloaderCommonPkg/Include/Guid/DeviceTableHobGuid.h [hunk 15]
```diff
@@ -0,0 +1,47 @@
...
+/// Device table Hob GUID
+///
+extern EFI_GUID gDeviceTableHobGuid;
+
+typedef struct {
+  UINT8        PciFunctionNumber;
+  UINT8        PciDeviceNumber;
+  UINT8        PciBusNumber;
+  UINT8        IsMmioDevice;
+} SBL_PCI_DEVICE;
```
[the remark this answers is not in the document]
Comment (gdong1) [reply BootloaderCommonPkg/Include/Guid/DeviceTableHobGuid.h]: will change to PLT_PCI_DEVICE

File: BootloaderCommonPkg/Include/Guid/DeviceTableHobGuid.h (hunk 10 above)
Comment (gdong1) [reply BootloaderCommonPkg/Include/Guid/DeviceTableHobGuid.h]: Signature is not used, will remove it.

File: BootloaderCommonPkg/Include/Guid/DeviceTableHobGuid.h (hunk 3 above)
Comment (gdong1) [reply BootloaderCommonPkg/Include/Guid/DeviceTableHobGuid.h]: already aligned, it is not required. and it is preferred method to avoid to use pack.

File: BootloaderCorePkg/Library/BootloaderLib/BootloaderLib.c [hunk 16]
```diff
@@ -352,3 +352,55 @@ GetFeatureCfg (
...
+  IN  BOOLEAN        PciMmioAddr
+  )
+{
+  SBL_DEVICE_TABLE   *DeviceTable;
+  SBL_DEVICE         *Device;
+  UINT32             DeviceBase;
+  UINT32             Index;
+
+  DeviceBase = 0;
+  DeviceTable = ((LOADER_GLOBAL_DATA *)GetLoaderGlobalDataPointer())->DeviceTable;
```
[the remark this answers is not in the document]
Comment (gdong1) [reply BootloaderCorePkg/Library/BootloaderLib/BootloaderLib.c]: OK. let me add GetDeviceTable().

File: BootloaderCorePkg/Library/BootloaderLib/BootloaderLib.c (hunk 11 above)
Comment (gdong1) [reply BootloaderCorePkg/Library/BootloaderLib/BootloaderLib.c]: OK. will add another Macro to convert to PCI base address.

File: Platform/ApollolakeBoardPkg/Library/Stage1BBoardInitLib/Stage1BBoardInitLib.c (hunk 4 above)
Comment (gdong1) [reply Platform/ApollolakeBoardPkg/Library/Stage1BBoardInitLib/Stage1BBoardInitLib.c]: good idea.

File: PayloadPkg/Library/SblParameterLib/SblParameterLib.c (hunk 13 above)
Comment (gdong1) [reply PayloadPkg/Library/SblParameterLib/SblParameterLib.c]: required specific for Android OS.

File: PayloadPkg/Library/PayloadLib/PayloadLib.c (hunk 12 above)
Comment (gdong1) [reply PayloadPkg/Library/PayloadLib/PayloadLib.c]: OK. will update it.

File: PayloadPkg/Library/PayloadLib/PayloadLib.c (hunk 9 above)
Comment (gdong1) [reply PayloadPkg/Library/PayloadLib/PayloadLib.c]: OK. will update it.

File: BootloaderCorePkg/Stage2/Stage2Support.c (hunk 8 above)
Comment (gdong1) [reply BootloaderCorePkg/Stage2/Stage2Support.c]: OK. will change it in another PR.

File: BootloaderCorePkg/Stage2/Stage2Support.c (hunk 7 above)
Comment (gdong1) [reply BootloaderCorePkg/Stage2/Stage2Support.c]: yes. we should create HOB even device number is 0 since payload will get this hob, and could know the device number is 0 from in the table header.

File: BootloaderCorePkg/Stage1B/Stage1B.c (hunk 6 above)
Comment (gdong1) [reply BootloaderCorePkg/Stage1B/Stage1B.c]: Device table should not be empty since boot option and others will depend on it.  And I don't see any issue to allocate memory if device number is really 0.

File: Platform/ApollolakeBoardPkg/Library/Stage1BBoardInitLib/Stage1BBoardInitLib.c (hunk 5 above)
Comment (gdong1) [reply Platform/ApollolakeBoardPkg/Library/Stage1BBoardInitLib/Stage1BBoardInitLib.c]: Will refine the structure and use a new method to avoid a new macro.
